src/codecs/openexr.rs

```diff
@@ -0,0 +1,185 @@
+                        .find(|chan| chan.name.eq(required)).is_some() // TODO eq_lowercase only if exrs supports it
+                );
+
+                !header.deep && has_rgb
+            })
+            .ok_or_else(|| ImageError::Decoding(DecodingError::new(
+                ImageFormatHint::Exact(ImageFormat::Exr),
+                "image does not contain non-deep rgb channels"
+            )))?;
+
+        Ok(Self { exr_reader, header_index })
+    }
+}
+
+
+impl<'a, R: 'a + BufRead + Seek> ImageDecoder<'a> for ExrDecoder<R> {
+    type Reader = Cursor<Vec<u8>>;
+
+    fn dimensions(&self) -> (u32, u32) {
+        let size = self.exr_reader.headers()[self.header_index].layer_size;
+        (size.width() as u32, size.height() as u32)
+    }
+
+    // TODO fn original_color_type(&self) -> ExtendedColorType {}
+
+    fn color_type(&self) -> ColorType {
+        // TODO adapt to actual exr color type
+        // let _channels = &self.exr_reader.headers()[self.header_index].channels;
+        ColorType::Rgba32F
+    }
+
+    /// Use `read_image` if possible,
+    /// as this method creates a whole new buffer to contain the entire image.
+    fn into_reader(self) -> ImageResult<Self::Reader> {
+        Ok(Cursor::new(image::decoder_to_vec(self)?)) // TODO no vec?
+    }
+
+    fn scanline_bytes(&self) -> u64 {
+        // we cannot always read individual scan lines for every file,
+        // as the tiles or lines in the file could be in random or reversed order.
+        // therefore we currently read all lines at once
+        // Todo: optimize for specific exr.line_order?
+       self.total_bytes()
+    }
+
+    fn read_image_with_progress<F: Fn(Progress)>(self, target: &mut [u8], progress_callback: F) -> ImageResult<()> {
+        let blocks_in_header = self.exr_reader.headers()[self.header_index].chunk_count as u64;
+        let height = self.dimensions().1 as usize;
+
+        let result = exr::prelude::read()
+            .no_deep_data().largest_resolution_level()
+            .rgba_channels(
+                |size, _channels|
+                    (size, vec![0_f32; size.area()*4])
+                ,
+
+                |(size, buffer), mut position, (r,g,b,a_or_1): (f32,f32,f32,f32)| {
+                    // TODO white point chromaticities + srgb/linear conversion?
+                    position.1 = height - (position.1 + 1); // openexr stores +y upwards, but we need +y downwards
+                    let first_f32_index = position.flat_index_for_size(*size);
+                    buffer[first_f32_index*4 .. (first_f32_index + 1)*4].copy_from_slice(&[r, g, b, a_or_1]);
+                }
+            )
+            .first_valid_layer() // TODO select exact layer by self.header_index?
+            .all_attributes()
+            .on_progress(move |progress| {
+                let mut prog = progress_callback;
+                prog(Progress::new((progress*blocks_in_header as f64) as u64, blocks_in_header)); // TODO precision errors?
```

Comment:
The `on_progress` function call does not compile, because it expects a function pointer, but a closure is given. But the very definition of the called function asks for a closure, and not for a function pointer?!

Definition:

```rust
/// Specify a function to be called regularly throughout the loading process.
/// Replaces all previously specified progress functions in this reader.
pub fn on_progress(self, on_progress: F) -> Self where F: FnMut(f64) { Self { on_progress, ..self } }
```

Call:
```rust
.on_progress(move |progress| {
    progress_callback(Progress::new(
        (progress*blocks_in_header as f64) as u64, blocks_in_header) // TODO precision errors?
    );
})
```

Error: 
```
error[E0308]: mismatched types
   --> src\codecs\openexr.rs:115:26
    |
115 |               .on_progress(move |progress| {
    |  __________________________^
116 | |                 progress_callback(Progress::new(
117 | |                     (progress*blocks_in_header as f64) as u64, blocks_in_header) // TODO precision errors?
118 | |                 );
119 | |             })
    | |_____________^ expected fn pointer, found closure
    |
    = note: expected fn pointer `fn(f64)`
                  found closure `[closure@src\codecs\openexr.rs:115:26: 119:14]`
```

Why does the compiler say it requires a function pointer when in fact it should indeed accept any closure? Any ideas?

Reply:
That is because the type  `F` mentioned here is the type parameter of [`ReadImage`](https://docs.rs/exr/1.2.0/exr/image/read/image/struct.ReadImage.html) and not one that you choose in `on_progress`. And the `all_attributes` method chooses to return specifically a `ReadImage<fn(_: f64), Self>`.

I'd say that this is something that the `exr` library should improve on.

Reply:
The function must have a separate type parameter, replacing the struct type parameter.